### PR TITLE
Add miniscope utils

### PR DESCRIPTION
**Pull request by weiglszonja**
Adds utility methods for Miniscope data
- `add_miniscope_device` adds Miniscope to an NWBFile based on provided metadata (can be microscope or behavioral camera settings)
- `add_miniscope_image_series` adds ImageSeries to an NWBFile in external mode 
- update usage in README

### CODE REVIEW

File: src/pynwb/ndx_miniscope/utils.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    Parameters
+    ----------
+    folder_path : str
+        The folder path that points to the main Miniscope folder.
+        The configuration files are expected to be located in subfolders within the main folder.
+        The configuration file should contain 'recordingStartTime' field that corresponds
+        to the start time of the Miniscope recordings.
+
+    """
+    folder_path = Path(folder_path)
```
[the remark this answers is not in the document]
Comment (weiglszonja) [reply src/pynwb/ndx_miniscope/utils.py:28]: possible: add version name to these utils + the base function name

File: src/pynwb/ndx_miniscope/miniscope.py [hunk 2]
```diff
@@ -1,60 +1,68 @@
...
     Returns
     -------
-    numpy.ndarray list if times in seconds
+    numpy.ndarray list of times in seconds
 
     """
-    if not fpath[-4:] == ".dat":
-        fpath = os.path.join(fpath, "timestamp.dat")
+    miniscope_version = get_miniscope_version(folder_path=folder_path)
+    if miniscope_version == version.Version("v4"):
```
Comment (CodyCBakerPhD) [inline src/pynwb/ndx_miniscope/miniscope.py:36]: Does the `packaging.version.Version` class handle the `"v"` in the name? I usually strip that out

File: src/pynwb/ndx_miniscope/miniscope.py
[the remark this answers is not in the document]
Comment (weiglszonja) [reply src/pynwb/ndx_miniscope/miniscope.py:1]: perfect, thanks I'll reorganise them!

File: README.md [hunk 3]
```diff
@@ -28,65 +50,95 @@ cd ndx-miniscope
...
-    nwb.add_acquisition(annotations)
+# Loading the timestamps
+behavcam_timestamps = get_timestamps(folder_path=folder_path, file_pattern="BehavCam*/timeStamps.csv")
+# Load the starting frames of the video files
+# Note this function requires to have `cv2` installed
+starting_frames = get_starting_frames(folder_path=folder_path, video_file_pattern="*/BehavCam*/*.avi")
 
-ms_files = natsorted(glob(os.path.join(data_dir, 'msCam*.avi')))
-behav_files = natsorted(glob(os.path.join(data_dir, 'behavCam*.avi')))
+# TODO: redirect to MiniscopeConverter
```
Comment (weiglszonja) [inline README.md:103]: Basically this part should be about how to add the `OnePhotonSeries` and `ImageSeries` to nwb which could be a redirect to the converter in `neuroconv`?

Comment (CodyCBakerPhD) [reply README.md:103]: Yeah, that sounds good for the README!

File: README.md [hunk 4]
```diff
@@ -4,14 +4,37 @@ This is a Neurodata Extension (NDX) for Neurodata Without Borders (NWB) 2.0 for
...
+        ├── 15_03_28/ (subfolder corresponding to the recording time)
+        │   ├── Miniscope/ (subfolder containing the microscope video stream)
+        │   │   ├── 0.avi (file containing the microscope video)
+        │   │   ├── metaData.json (file containing the metadata for the microscope device)
+        │   │   └── timeStamps.csv (file containing the timing of this video stream)
+        │   ├── BehavCam_2/ (subfolder containing the behavioral video stream)
+        │   │   ├── 0.avi (file containing the bevavioral video)
+        │   │   ├── metaData.json (file containing the metadata for the behavioral camera)
+        │   │   └── timeStamps.csv (file containing the timing of this video stream)
+        │   └── metaData.json (file containing metadata for the recording, such as the start time)
```
Comment (bendichter) [inline README.md:23]:
```suggestion
        │   │   ├── 0.avi (microscope video)
        │   │   ├── metaData.json (metadata for the microscope device)
        │   │   └── timeStamps.csv (timing of this video stream)
        │   ├── BehavCam_2/ (subfolder containing the behavioral video stream)
        │   │   ├── 0.avi (bevavioral video)
        │   │   ├── metaData.json (metadata for the behavioral camera)
        │   │   └── timeStamps.csv (timing of this video stream)
        │   └── metaData.json (metadata for the recording, such as the start time)
```

it feels to me that it's self-evident these are files

File: src/pynwb/ndx_miniscope/utils/nwb.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+def add_miniscope_device(nwbfile: NWBFile, device_metadata: Dict) -> NWBFile:
+    """
+    Adds a Miniscope device based on provided metadata.
+    Can be used to add device for the microscope and the behavioral camera.
+
+    Parameters
+    ----------
+    nwbfile : NWBFile
+        The nwbfile to add the Miniscope device to.
+    device_metadata: Dict
```
Comment (bendichter) [inline src/pynwb/ndx_miniscope/utils/nwb.py:21]:
```suggestion
    device_metadata: dict
```

File: src/pynwb/ndx_miniscope/utils/nwb.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+    device = Miniscope(**device_metadata_copy)
+    nwbfile.add_device(device)
+
+    return nwbfile
+
+
+def add_miniscope_image_series(
+    nwbfile: NWBFile,
+    metadata: Dict,
```
Comment (bendichter) [inline src/pynwb/ndx_miniscope/utils/nwb.py:48]:
```suggestion
    metadata: dict,
```

File: src/pynwb/ndx_miniscope/utils/settings.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        return version.Version("4")
+
+    legacy_config_files = list(folder_path.rglob("*.dat"))
+    if legacy_config_files:
+        return version.Version("3")
+
+    raise NotImplementedError("Could not determine the version of the Miniscope.")
+
+
+def read_miniscope_config(folder_path: str) -> Dict:
```
Comment (bendichter) [inline src/pynwb/ndx_miniscope/utils/settings.py:47]:
```suggestion
def read_miniscope_config(folder_path: str) -> dict:
```

File: src/pynwb/ndx_miniscope/utils/settings.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    Parameters
+    ----------
+    folder_path: str
+        The folder path where the settings of the microscope or the behavioral camera is located.
+
+
+    Returns
+    -------
+    Dict
```
Comment (bendichter) [inline src/pynwb/ndx_miniscope/utils/settings.py:67]:
```suggestion
    dict
```

File: README.md [hunk 9]
```diff
@@ -28,65 +51,95 @@ cd ndx-miniscope
...
-    nwb.add_acquisition(annotations)
+# Loading the timestamps
+behavcam_timestamps = get_timestamps(folder_path=folder_path, file_pattern="BehavCam*/timeStamps.csv")
+# Load the starting frames of the video files
+# Note this function requires to have `cv2` installed
+starting_frames = get_starting_frames(folder_path=folder_path, video_file_pattern="*/BehavCam*/*.avi")
 
-ms_files = natsorted(glob(os.path.join(data_dir, 'msCam*.avi')))
-behav_files = natsorted(glob(os.path.join(data_dir, 'behavCam*.avi')))
+# TODO: redirect to MiniscopeConverter
```
Comment (CodyCBakerPhD) [inline README.md:104]:
```suggestion
```

File: src/pynwb/ndx_miniscope/utils/settings.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    df = pd.read_csv(fpath, sep="\t").loc[0]
+
+    return dict(
+        name="Miniscope",
+        excitation=int(df["excitation"]),
+        msCamExposure=int(df["msCamExposure"]),
+    )
+
+
+def read_v4_miniscope_config(folder_path: str) -> Dict:
```
Comment (bendichter) [inline src/pynwb/ndx_miniscope/utils/settings.py:89]:
```suggestion
def read_v4_miniscope_config(folder_path: str) -> dict:
```

File: src/pynwb/ndx_miniscope/utils/nwb.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
+from copy import deepcopy
+from typing import Dict, List, Optional
```
Comment (CodyCBakerPhD) [inline src/pynwb/ndx_miniscope/utils/nwb.py:2]:
```suggestion
from typing import List, Optional
```

File: src/pynwb/ndx_miniscope/utils/settings.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
+import json
+import os
+from pathlib import Path
+from typing import Dict
```
Comment (CodyCBakerPhD) [inline src/pynwb/ndx_miniscope/utils/settings.py:4]:
```suggestion
```

**Comment by CodyCBakerPhD** [comment]
@weiglszonja I added a reminder note to your to do list to come back and update the write step once the NeuroConv stuff is in place

**Pull request merged**